### PR TITLE
Various improvements

### LINKED ISSUE #3: `linux-static-hidraw` doesn't detect coldcard
Hey there, I'm building a wallet over at https://github.com/cloudhead/nakamoto and trying to integrate coldcard. With the default hidapi features, my CC was not detected. However, with this change, it was:

```diff
diff --git a/coldcard/Cargo.toml b/coldcard/Cargo.toml
index b7f9252..b93037f 100644
--- a/coldcard/Cargo.toml
+++ b/coldcard/Cargo.toml
@@ -13,5 +13,5 @@ categories = ["command-line-utilities", "cryptography::cryptocurrencies", "hardw
 aes-ctr = "0.6.0"
 base58 = "0.2.0"
 enum-as-inner = "0.5.1"
-hidapi = { version = "1.4.2", default-features = false, features = ["linux-static-hidraw"] }
+hidapi = { version = "1.4.2", default-features = false, features = ["linux-static-libusb"] }
```

I think it would make sense to either enable both, or to expose those features so that they can be enabled when depending on this crate. What do you think?

### PR DESCRIPTION
**Pull request by cloudhead**
Hey there, I've been using a fork of your library but thought I would propose the changes here, since it might benefit more people. Each commit should describe the change.

* We should eventually also replace `aes-ctr` with `aes`, since the former is deprecated.
* I'd eventually recommend running clippy more fully, as I just made some more obvious lint fixes.

Closes #3

### CODE REVIEW

**Review by alfred-hodler — CHANGES_REQUESTED** [review]
Thank you for this PR! Almost everything looks solid, I just have a few minor points that I'd like addressed, including one potential regression.

Otherwise, this is almost good to go :rocket: 

If you'd like an immediate crates.io release, please bump the patch portion of the version.

**Comment by alfred-hodler** [comment]
> We should eventually also replace aes-ctr with aes, since the former is deprecated.

Agreed. I picked an older version simply because the API was cleaner and easier to use, but this change needs to be made for security reasons.

File: coldcard/src/firmware.rs [hunk 1]
```diff
@@ -35,7 +35,7 @@ impl Firmware {
             file.read_exact(&mut two_u32)?;
             let elements = decode_u32(two_u32.get(4..8))?;
 
-            for _ in 0..elements {
+            if (0..elements).next().is_some() {
```
[the remark this answers is not in the document]
Comment (cloudhead) [reply coldcard/src/firmware.rs:38]: I was also confused at first, but if you look at the loop, there is a return statement at the end (and no `continue` statement), so it only loops once. Perhaps this points to a bug in the code?

File: coldcard/Cargo.toml [hunk 2]
```diff
@@ -9,9 +9,12 @@ description = "Coldcard Wallet Interface Library in Rust"
 keywords = ["coldcard", "bitcoin", "wallet"]
 categories = ["command-line-utilities", "cryptography::cryptocurrencies", "hardware-support"]
 
+[features]
+default = ["hidapi/default"]
+linux-static-hidraw = ["hidapi/linux-static-hidraw"]
+linux-static-libusb = ["hidapi/linux-static-libusb"]
```
[the remark this answers is not in the document]
Comment (cloudhead) [reply coldcard/Cargo.toml:15]: Yeah, can do.

File: coldcard/Cargo.toml [hunk 3]
```diff
@@ -9,9 +9,12 @@ description = "Coldcard Wallet Interface Library in Rust"
...
+
 [dependencies]
 aes-ctr = "0.6.0"
-base58 = "0.2.0"
-enum-as-inner = "0.5.1"
-hidapi = { version = "1.4.2", default-features = false, features = ["linux-static-hidraw"] }
-secp256k1 = { version = "0.24.0", features = ["rand-std", "bitcoin_hashes"] }
+bitcoin = { version = "0.29.1", features = ["rand"] }
```
[the remark this answers is not in the document]
Comment (cloudhead) [reply coldcard/Cargo.toml:19]: I generally agree, but here's why I made this change:
1. We get to drop the `base58` dependency and use the one in `bitcoin`
2. It's almost a given that a crate importing `coldcard` is also already depending on `bitcoin`, for all the wallet-related types in there.
3. There are some places in `coldcard` that could benefit from using `bitcoin` crate types, eg. `DerivationPath` and other `bitcoin::util::bip32` functions, as well as the `Address` type which should ideally be returned from the `Coldcard::address` method. I'm sure I will encounter others.

If you don't agree with the above, I would suggest perhaps making it an optional dependency that can be enabled for conversion methods and dependency consolidation.

**Comment by cloudhead** [comment]
> Agreed. I picked an older version simply because the API was cleaner and easier to use, but this change needs to be made for security reasons.

Yeah, I tried moving it over to `aes` but there were too many differences, so I think it will take a bit more work.

File: coldcard/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (alfred-hodler) [reply coldcard/Cargo.toml:19]: The third reason is particularly compelling. We can get rid of those custom structs by using `rust-bitcoin` types in those places.

File: coldcard/src/firmware.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alfred-hodler) [reply coldcard/src/firmware.rs:38]: Clippy did the right thing. I just checked Coinkite's Python library and it's also only looking at the first offset. It appears there should always be a firmware blob at the first offset, at least the way they build their `.dfu` files, so this is fine.

**Comment by alfred-hodler** [comment]
Looks good. We can replace those custom bip32 structs in a different unit of work, along with upgrading the AES crate.

**Comment by cloudhead** [comment]
Great!

**Pull request merged**